File: android/app/src/main/java/dev/pepotech/pepomote/control/ButtonState.kt
package dev.pepotech.pepomote.control

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import java.util.concurrent.CopyOnWriteArraySet
import java.util.concurrent.atomic.AtomicInteger

/**
 * Estado de botones compartido entre la UI (escribe) y el hilo de sensores (lee).
 * Bits según PROTOCOL.md §4.2. Las pulsaciones pasan por [PressLatch]: un
 * toque corto se retiene para darle más oportunidades de llegar al receptor.
 */
object ButtonState {
    const val A = 1 shl 0
    const val B = 1 shl 1
    const val DPAD_UP = 1 shl 2
    const val DPAD_DOWN = 1 shl 3
    const val DPAD_LEFT = 1 shl 4
    const val DPAD_RIGHT = 1 shl 5
    const val PLUS = 1 shl 6
    const val MINUS = 1 shl 7
    const val HOME = 1 shl 8
    const val ONE = 1 shl 9
    const val TWO = 1 shl 10
    const val MEDIA_VOL_UP = 1 shl 11
    const val MEDIA_VOL_DOWN = 1 shl 12
    const val MEDIA_MUTE = 1 shl 13
    const val MEDIA_PLAY_PAUSE = 1 shl 14
    const val MEDIA_NEXT = 1 shl 15
    const val MEDIA_PREV = 1 shl 16

    /** Nunchuk (móvil de la otra mano). */
    const val C = 1 shl 17
    const val Z = 1 shl 18

    /** Wii U GamePad / Pro Controller (modo Cemu). */
    const val X = 1 shl 19
    const val Y = 1 shl 20
    const val L = 1 shl 21
    const val R = 1 shl 22
    const val ZL = 1 shl 23
    const val ZR = 1 shl 24

    /** Click del stick izquierdo / derecho (L3 / R3). */
    const val STICK_L = 1 shl 25
    const val STICK_R = 1 shl 26

    /** Soplar al micrófono del GamePad. */
    const val MIC = 1 shl 27

    /** Wii U: cambiar TV ↔ GamePad. Switch: Capturar. */
    const val SCREEN = 1 shl 28

    /** Precisión (modo puntero): mientras se mantiene, el cursor va al 40 %. */
    const val PRECISION = 1 shl 29

    /**
     * Acercar (modo Dolphin): mientras se mantiene, el receptor acerca el
     * Mando de Wii emulado a la pantalla (juegos que piden acercar el mando).
     */
    const val NEAR = 1 shl 30

    /**
     * Dedo en la pantalla táctil del GamePad: instantánea inmutable, así el
     * hilo de sensores lee x, y y `down` coherentes entre sí de una sola vez.
     * x, y = fracción de la pantalla en 0..65535, origen arriba-izquierda.
     */
    class Touch(val x: Int, val y: Int, val down: Boolean)

    private val NO_TOUCH = Touch(0, 0, false)

    /** A reset also invalidates unsent edges when leaving a screen or changing mode. */
    internal class Change(val buttons: Int, val reset: Boolean = false)

    private val stateLock = Any()
    private val mask = AtomicInteger(0)
    private val buttonChangeListeners = CopyOnWriteArraySet<(Change) -> Unit>()
    private val recenter = AtomicInteger(0)
    private val scrollAcc = AtomicInteger(0)

    /** Stick del Nunchuk o izquierdo del GamePad, −127..127, +x derecha, +y arriba. */
    private val stickXv = AtomicInteger(0)
    private val stickYv = AtomicInteger(0)

    /** Stick derecho del GamePad, misma convención. */
    private val stick2Xv = AtomicInteger(0)
    private val stick2Yv = AtomicInteger(0)

    @Volatile
    private var touchV: Touch = NO_TOUCH

    private val handler = Handler(Looper.getMainLooper())
    private val latch = PressLatch(object : PressLatch.Scheduler {
        override fun now(): Long = SystemClock.uptimeMillis()
        override fun postDelayed(delayMs: Long, task: Runnable) {
            handler.postDelayed(task, delayMs)
        }

        override fun cancel(task: Runnable) = handler.removeCallbacks(task)
    }) { bit, down ->
        // PressLatch already holds its lock. Mutation and notification are one
        // ordered operation for every subscriber, including a concurrent reset.
        synchronized(stateLock) {
            val previous = mask.get()
            val next = if (down) previous or bit else previous and bit.inv()
            if (next != previous) {
                mask.set(next)
                val change = Change(next)
                buttonChangeListeners.forEach { it(change) }
            }
        }
    }

    /** Aplica la pulsación al estado y avisa al emisor sin esperar al siguiente sensor. */
    fun set(bit: Int, down: Boolean) = latch.set(bit, down)

    fun current(): Int = mask.get()

    /**
     * Snapshot inicial y cambios posteriores forman una sola secuencia ordenada.
     * El listener solo encola trabajo: no toma el lock del emisor ni modifica botones.
     */
    internal fun addButtonChangeListener(listener: (Change) -> Unit) {
        synchronized(stateLock) {
            if (buttonChangeListeners.add(listener)) listener(Change(mask.get()))
        }
    }

    internal fun removeButtonChangeListener(listener: (Change) -> Unit) {
        synchronized(stateLock) { buttonChangeListeners.remove(listener) }
    }

    fun bumpRecenter() {
        recenter.incrementAndGet()
    }

    fun recenterCount(): Int = recenter.get() and 0xFF

    fun addScroll(dyPx: Int) {
        scrollAcc.addAndGet(dyPx)
    }

    /** Vacía el acumulador de scroll (lo llama el hilo de sensores por paquete). */
    fun drainScroll(): Int = scrollAcc.getAndSet(0).coerceIn(-32768, 32767)

    /** Posición del stick (el stick va sin latch: es continuo, no un flanco). */
    fun setStick(x: Int, y: Int) {
        stickXv.set(x.coerceIn(-127, 127))
        stickYv.set(y.coerceIn(-127, 127))
    }

    fun stickX(): Int = stickXv.get()
    fun stickY(): Int = stickYv.get()

    /** Stick derecho del GamePad (solo lo escribe la pantalla GamePad). */
    fun setStick2(x: Int, y: Int) {
        stick2Xv.set(x.coerceIn(-127, 127))
        stick2Yv.set(y.coerceIn(-127, 127))
    }

    fun stickRX(): Int = stick2Xv.get()
    fun stickRY(): Int = stick2Yv.get()

    /**
     * Pantalla táctil del GamePad: `x`, `y` en 0..65535 (se recortan) y
     * `down` = hay dedo. Al soltar se conservan las últimas coordenadas
     * (al receptor le da igual: sin FLAG_TOUCH no las mira).
     */
    fun setTouch(x: Int, y: Int, down: Boolean) {
        touchV = Touch(x.coerceIn(0, 65535), y.coerceIn(0, 65535), down)
    }

    /** Instantánea atómica del táctil. */
    fun touch(): Touch = touchV

    /** Todo suelto: botones, scroll, ambos sticks y táctil. */
    fun reset() {
        // Match set/onWire's lock order. A new press cannot slip between
        // cancelling the old latch and publishing its reset notification.
        synchronized(latch) {
            latch.reset()
            synchronized(stateLock) {
                mask.set(0)
                scrollAcc.set(0)
                stickXv.set(0)
                stickYv.set(0)
                stick2Xv.set(0)
                stick2Yv.set(0)
                touchV = NO_TOUCH
                val change = Change(0, reset = true)
                buttonChangeListeners.forEach { it(change) }
            }
        }
    }
}

File: ios/PepoMote/Control/ButtonState.swift
import Foundation

/// Bits de botones (PROTOCOL.md §4.2). Los mismos que Android y el receptor.
enum Btn {
    static let a: UInt32 = 1 << 0
    static let b: UInt32 = 1 << 1
    static let dpadUp: UInt32 = 1 << 2
    static let dpadDown: UInt32 = 1 << 3
    static let dpadLeft: UInt32 = 1 << 4
    static let dpadRight: UInt32 = 1 << 5
    static let plus: UInt32 = 1 << 6
    static let minus: UInt32 = 1 << 7
    static let home: UInt32 = 1 << 8
    static let one: UInt32 = 1 << 9
    static let two: UInt32 = 1 << 10
    static let mediaVolUp: UInt32 = 1 << 11
    static let mediaVolDown: UInt32 = 1 << 12
    static let mediaMute: UInt32 = 1 << 13
    static let mediaPlayPause: UInt32 = 1 << 14
    static let mediaNext: UInt32 = 1 << 15
    static let mediaPrev: UInt32 = 1 << 16
    /// Nunchuk (móvil de la otra mano).
    static let c: UInt32 = 1 << 17
    static let z: UInt32 = 1 << 18
    /// Wii U GamePad / Pro Controller (modo Cemu).
    static let x: UInt32 = 1 << 19
    static let y: UInt32 = 1 << 20
    static let l: UInt32 = 1 << 21
    static let r: UInt32 = 1 << 22
    static let zl: UInt32 = 1 << 23
    static let zr: UInt32 = 1 << 24
    /// Click del stick izquierdo / derecho (L3 / R3).
    static let stickL: UInt32 = 1 << 25
    static let stickR: UInt32 = 1 << 26
    /// Soplar al micrófono del GamePad.
    static let mic: UInt32 = 1 << 27
    /// Cambiar la vista TV ↔ pantalla del GamePad (función de Cemu).
    static let screen: UInt32 = 1 << 28
    /// Precisión (modo puntero): mientras se mantiene, el cursor va al 40 %.
    static let precision: UInt32 = 1 << 29
}

/// Reloj + temporizador del latch (cola principal en la app; virtual en tests).
protocol LatchScheduler {
    /// Milisegundos monótonos.
    func now() -> Int64
    /// Ejecuta `block` pasados `afterMs` ms, salvo que la tarea se cancele antes.
    func schedule(afterMs: Int64, _ block: @escaping () -> Void) -> LatchTask
}

/// Tarea programada cancelable.
final class LatchTask {
    private(set) var cancelled = false
    func cancel() { cancelled = true }
}

/// Temporizadores en la cola principal (la UI escribe los botones desde ahí).
final class MainQueueScheduler: LatchScheduler {
    func now() -> Int64 {
        Int64(DispatchTime.now().uptimeNanoseconds / 1_000_000)
    }

    func schedule(afterMs: Int64, _ block: @escaping () -> Void) -> LatchTask {
        let task = LatchTask()
        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(Int(max(0, afterMs)))) {
            if !task.cancelled { block() }
        }
        return task
    }
}

/// Estado de botones compartido entre la UI (escribe) y el hilo de sensores
/// (lee). Las pulsaciones pasan por `PressLatch`: un toque, por corto que sea,
/// dura lo suficiente en el cable para llegar.
final class ButtonState {
    static let shared = ButtonState()

    /// Dedo en la pantalla táctil del GamePad: instantánea inmutable (x, y en
    /// 0..65535, origen arriba-izquierda).
    struct Touch {
        let x: Int
        let y: Int
        let down: Bool
    }

    private let lock = NSLock()
    private var mask: UInt32 = 0
    private var recenter: Int = 0
    private var scrollAcc: Int = 0
    private var stickV = (x: 0, y: 0)
    private var stick2V = (x: 0, y: 0)
    private var touchV = Touch(x: 0, y: 0, down: false)
    private var latch: PressLatch!

    init(scheduler: LatchScheduler = MainQueueScheduler()) {
        latch = PressLatch(scheduler: scheduler) { [weak self] bit, down in
            self?.wire(bit, down)
        }
    }

    private func wire(_ bit: UInt32, _ down: Bool) {
        lock.lock()
        if down { mask |= bit } else { mask &= ~bit }
        lock.unlock()
    }

    /// Pulsación física/táctil: el flanco de bajada sale al instante.
    func set(_ bit: UInt32, _ down: Bool) {
        latch.set(bit, down)
    }

    func current() -> UInt32 {
        lock.lock(); defer { lock.unlock() }
        return mask
    }

    func bumpRecenter() {
        lock.lock(); recenter &+= 1; lock.unlock()
    }

    func recenterCount() -> Int {
        lock.lock(); defer { lock.unlock() }
        return recenter & 0xFF
    }

    func addScroll(_ dyPx: Int) {
        lock.lock(); scrollAcc += dyPx; lock.unlock()
    }

    /// Vacía el acumulador de scroll (lo llama el hilo de sensores por paquete).
    func drainScroll() -> Int {
        lock.lock(); defer { lock.unlock() }
        let v = scrollAcc
        scrollAcc = 0
        return min(max(v, -32768), 32767)
    }

    /// Stick del Nunchuk o izquierdo del GamePad, −127..127, +x derecha, +y arriba.
    func setStick(_ x: Int, _ y: Int) {
        lock.lock(); stickV = (Self.clamp127(x), Self.clamp127(y)); lock.unlock()
    }

    func stickX() -> Int { lock.lock(); defer { lock.unlock() }; return stickV.x }
    func stickY() -> Int { lock.lock(); defer { lock.unlock() }; return stickV.y }

    /// Stick derecho del GamePad (solo lo escribe la pantalla GamePad).
    func setStick2(_ x: Int, _ y: Int) {
        lock.lock(); stick2V = (Self.clamp127(x), Self.clamp127(y)); lock.unlock()
    }

    func stickRX() -> Int { lock.lock(); defer { lock.unlock() }; return stick2V.x }
    func stickRY() -> Int { lock.lock(); defer { lock.unlock() }; return stick2V.y }

    /// Pantalla táctil del GamePad: x, y en 0..65535 (se recortan) y `down` = hay dedo.
    func setTouch(_ x: Int, _ y: Int, _ down: Bool) {
        lock.lock()
        touchV = Touch(x: min(max(x, 0), 65535), y: min(max(y, 0), 65535), down: down)
        lock.unlock()
    }

    func touch() -> Touch {
        lock.lock(); defer { lock.unlock() }
        return touchV
    }

    /// Todo suelto: botones, scroll, ambos sticks y táctil.
    func reset() {
        latch.reset()
        lock.lock()
        mask = 0
        scrollAcc = 0
        stickV = (0, 0)
        stick2V = (0, 0)
        touchV = Touch(x: 0, y: 0, down: false)
        lock.unlock()
    }

    private static func clamp127(_ v: Int) -> Int { min(max(v, -127), 127) }
}
